File: Vanukov-ML-site/simulator/simulator.js
import { WebSocketServer } from 'ws';
import * as fs from 'fs';
import { dirname, join } from 'path';
import {parse} from 'csv-parse'
import { fileURLToPath } from 'url';

//init server
const wss = new WebSocketServer({ port: 5001 });
console.log('WebSocket server is running on ws://0.0.0.0:5001');

// Получаем __dirname в ES Modules
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Column translations
const columnTranslations = {
    'номер измерения': 'Measurement ID',
    'дата': 'Date',
    'давление КВС, точка1': 'blast furnace pressure, point 1',
    'давление КВС, точка2': 'blast furnace pressure, point 2',
    'давление природный газ': 'natural gas pressure',
    'конвейер 31, производительность': 'conveyor 31, productivity',
    'конвейер 31, скорость': 'conveyor 31, speed',
    'конвейер 32, производительность': 'conveyor 32, productivity',
    'конвейер 32, скорость': 'conveyor 32, speed',
    'питатель1, уровень': 'feeder 1, level',
    'питатель1, скорость': 'feeder 1, speed',
    'питатель1, производительность': 'feeder 1, productivity',
    'питатель2, уровень': 'feeder 2, level',
    'питатель2, скорость': 'feeder 2, speed',
    'питатель2, производительность': 'feeder 2, productivity',
    'питатель3, уровень': 'feeder 3, level',
    'питатель3, скорость': 'feeder 3, speed',
    'питатель3, производительность': 'feeder 3, productivity',
    'питатель4, уровень': 'feeder 4, level',
    'питатель4, скорость': 'feeder 4, speed',
    'питатель4, производительность': 'feeder 4, productivity',
    'питатель5, уровень': 'feeder 5, level',
    'питатель5, скорость': 'feeder 5, speed',
    'питатель5, производительность': 'feeder 5, productivity',
    'питатель6, уровень': 'feeder 6, level',
    'питатель6, скорость': 'feeder 6, speed',
    'питатель6, производительность': 'feeder 6, productivity',
    'питатель7, скорость': 'feeder 7, speed',
    'питатель8, скорость': 'feeder 8, speed',
    'питатель7, уровень': 'feeder 7, level',
    'питатель8, уровень': 'feeder 8, level',
    'разрежение в аптейке': 'vacuum in the bunker',
    'расход КВС': 'blast furnace flow',
    'расход природного газа': 'natural gas flow',
    'содержание кислорода в КВС': 'oxygen content in the blast furnace',
    'температура КВС': 'blast furnace temperature',
    'температура отходящих газов в аптейке': 'temperature of outgoing gases in the bunker',
    'температура пода, шлаковый сифон': 'temperature of the feed, slag siphon',
    'температура пода, штейновый сифон': 'temperature of the feed, matte siphon',
    'температура пода, зона плавления, точка1': 'temperature of the feed, melting zone, point 1',
    'температура пода, зона плавления, точка2': 'temperature of the feed, melting zone, point 2',
    'температура природного газа': 'temperature of natural gas',
};

function loadCSV(filePath) {
    return new Promise((resolve, reject) => {
        const results = [];
        fs.createReadStream(filePath)
            .pipe(parse({ columns: true, skip_lines: 2 }))
            .on('data', (row) => results.push(row))
            .on('end', () => resolve(results))
            .on('error', (err) => reject(err));
    });
}

async function loadAndPreprocessData() {
    try {
        // Load CSV files
        const data1 = await loadCSV(join(__dirname, 'data/data1.csv'));
        const data2 = await loadCSV(join(__dirname, 'data/data2.csv'));
        const data3 = await loadCSV(join(__dirname, 'data/data3.csv'));

        // Rename columns
        const renameColumns = (data) => {
            return data.map(row => {
                const newRow = {};
                for (const key in row) {
                    newRow[columnTranslations[key] || key] = row[key];
                }
                return newRow;
            });
        };

        // Merge data
        let data = [...data1, ...data2, ...data3];
        data = renameColumns(data)
        // Convert Date and filter out invalid rows
        data = data
            .map(row => ({
                ...row,
                Date: new Date(row.Date).toString() !== 'Invalid Date' ? row.Date : null
            }))
            .filter(row => row.Date !== null);

        data = data.map(row => ({
        ...row,
        'Total charge rate, t/h': Number(row['conveyor 31, productivity']) + Number(row['conveyor 32, productivity']) || 0,
        'Temperature of feed in the smelting zone, °C': (
            (Number(row['temperature of the feed, melting zone, point 1']) || 0) +
            (Number(row['temperature of the feed, melting zone, point 2']) || 0)
        ) / 2,
        'Temperature of exhaust gases in the off-gas duct, °C': Number(row['temperature of outgoing gases in the bunker']) || 0,
        'Overall blast volume, m3/h': Number(row['blast furnace flow']) || 0,
        'Oxygen content in the blast (degree of oxygen enrichment in the blowing), %': Number(row['oxygen content in the blast furnace']) || 0,
        'feeder 2, speed': Number(row['feeder 2, speed']) || 0,
        }));

        return data;
    } catch (error) {
        console.error('Error in data preprocessing:', error);
        throw error;
  }
}

async function simulateFurnace(ws) {
    try {
        const data = await loadAndPreprocessData();
        for (const row of data) {
            if (ws.readyState === 1) { // WebSocket.OPEN
                ws.send(JSON.stringify(row));
                await new Promise(resolve => setTimeout(resolve, 1000)); // 1-second delay
            } else {
                console.log('WebSocket connection closed, stopping simulation');
                break;
            }
        }
    } catch (error) {
        console.error('Error in simulateFurnace:', error);
    }
}

// Handle WebSocket connections
wss.on('connection', (ws) => {
    console.log('New WebSocket connection');

    // Start simulation for this client
    simulateFurnace(ws);

    ws.on('close', () => {
        console.log('WebSocket connection closed');
    });

    ws.on('error', (error) => {
        console.error('WebSocket error:', error);
    });
});

File: Vanukov-ML-site/simulator/simulator.py
import asyncio
import websockets
import pandas as pd
import numpy as np
import os
from datetime import datetime

# Column translations
column_translations = {
    'номер измерения': 'Measurement ID',
    'дата': 'Date',
    'давление КВС, точка1': 'blast furnace pressure, point 1',
    'давление КВС, точка2': 'blast furnace pressure, point 2',
    'давление природный газ': 'natural gas pressure',
    'конвейер 31, производительность': 'conveyor 31, productivity',
    'конвейер 31, скорость': 'conveyor 31, speed',
    'конвейер 32, производительность': 'conveyor 32, productivity',
    'конвейер 32, скорость': 'conveyor 32, speed',
    'питатель1, уровень': 'feeder 1, level',
    'питатель1, скорость': 'feeder 1, speed',
    'питатель1, производительность': 'feeder 1, productivity',
    'питатель2, уровень': 'feeder 2, level',
    'питатель2, скорость': 'feeder 2, speed',
    'питатель2, производительность': 'feeder 2, productivity',
    'питатель3, уровень': 'feeder 3, level',
    'питатель3, скорость': 'feeder 3, speed',
    'питатель3, производительность': 'feeder 3, productivity',
    'питатель4, уровень': 'feeder 4, level',
    'питатель4, скорость': 'feeder 4, speed',
    'питатель4, производительность': 'feeder 4, productivity',
    'питатель5, уровень': 'feeder 5, level',
    'питатель5, скорость': 'feeder 5, speed',
    'питатель5, производительность': 'feeder 5, productivity',
    'питатель6, уровень': 'feeder 6, level',
    'питатель6, скорость': 'feeder 6, speed',
    'питатель6, производительность': 'feeder 6, productivity',
    'питатель7, скорость': 'feeder 7, speed',
    'питатель8, скорость': 'feeder 8, speed',
    'питатель7, уровень': 'feeder 7, level',
    'питатель8, уровень': 'feeder 8, level',
    'разрежение в аптейке': 'vacuum in the bunker',
    'расход КВС': 'blast furnace flow',
    'расход природного газа': 'natural gas flow',
    'содержание кислорода в КВС': 'oxygen content in the blast furnace',
    'температура КВС': 'blast furnace temperature',
    'температура отходящих газов в аптейке': 'temperature of outgoing gases in the bunker',
    'температура пода, шлаковый сифон': 'temperature of the feed, slag siphon',
    'температура пода, штейновый сифон': 'temperature of the feed, matte siphon',
    'температура пода, зона плавления, точка1': 'temperature of the feed, melting zone, point 1',
    'температура пода, зона плавления, точка2': 'temperature of the feed, melting zone, point 2',
    'температура природного газа': 'temperature of natural gas',
}

async def load_and_preprocess_data():
    try:
        # Load CSV files
        data_files = ['data/data1.csv', 'data/data2.csv', 'data/data3.csv']
        dataframes = []
        for file in data_files:
            if not os.path.exists(file):
                print(f"Error: File {file} not found")
                continue
            print(f"Loading file: {file}")
            df = pd.read_csv(file)
            print(f"Loaded {len(df)} rows from {file}")
            print(f"Columns in {file}: {list(df.columns)}")
            dataframes.append(df)
        
        if not dataframes:
            print("Error: No CSV files loaded")
            return []

        data = pd.concat(dataframes, ignore_index=True)
        print(f"Total rows after concatenation: {len(data)}")

        # Rename columns
        data = data.rename(columns=column_translations)
        print(f"Columns after renaming: {list(data.columns)}")

        # Handle Date column
        if 'Date' not in data.columns and 'дата' not in data.columns:
            print("Warning: 'Date' column not found, generating timestamps")
            data['Date'] = [datetime.now().isoformat() for _ in range(len(data))]
        else:
            data['Date'] = pd.to_datetime(data.get('Date', data.get('дата')), errors='coerce')
            # If Date is invalid, generate timestamps
            if data['Date'].isna().all():
                print("Warning: All 'Date' values are invalid, generating timestamps")
                data['Date'] = [datetime.now().isoformat() for _ in range(len(data))]
        
        # Filter out rows with invalid Date
        data = data.dropna(subset=['Date'])
        print(f"Rows after filtering invalid dates: {len(data)}")


        # Calculate derived features
        data['Total charge rate, t/h'] = data['conveyor 31, productivity'].astype(float) + data['conveyor 32, productivity'].astype(float)
        data['Temperature of feed in the smelting zone, °C'] = (
            data['temperature of the feed, melting zone, point 1'].astype(float) +
            data['temperature of the feed, melting zone, point 2'].astype(float)
        ) / 2
        data['Temperature of exhaust gases in the off-gas duct, °C'] = data['temperature of outgoing gases in the bunker'].astype(float)
        data['Overall blast volume, m3/h'] = data['blast furnace flow'].astype(float)
        data['Oxygen content in the blast, %'] = data['oxygen content in the blast furnace'].astype(float)
        data['feeder 2, speed'] = data['feeder 2, speed'].astype(float)

        # Ensure numeric types for all relevant columns
        numeric_columns = [col for col in data.columns if col != 'Date' and col != 'Measurement ID']
        for col in numeric_columns:
            data[col] = pd.to_numeric(data[col], errors='coerce').fillna(0)

        return data.to_dict('records')
    except Exception as e:
        print(f"Error in data preprocessing: {e}")
        return []

async def simulate_furnace(websocket):
    try:
        data = await load_and_preprocess_data()
        if not data:
            print("No data available to simulate")
            return

        # Define noise ranges for dynamic simulation
        noise_ranges = {
            'blast furnace pressure, point 1': (-5, 5),
            'blast furnace pressure, point 2': (-5, 5),
            'natural gas pressure': (-0.05, 0.05),
            'conveyor 31, productivity': (-10, 10),
            'conveyor 31, speed': (-0.5, 0.5),
            'conveyor 32, productivity': (-10, 10),
            'conveyor 32, speed': (-0.5, 0.5),
            'feeder 1, level': (-5, 5),
            'feeder 1, speed': (-3, 3),
            'feeder 1, productivity': (-10, 10),
            'feeder 2, level': (-5, 5),
            'feeder 2, speed': (-3, 3),
            'feeder 2, productivity': (-10, 10),
            'feeder 3, level': (-5, 5),
            'feeder 3, speed': (-3, 3),
            'feeder 3, productivity': (-10, 10),
            'feeder 4, level': (-5, 5),
            'feeder 4, speed': (-3, 3),
            'feeder 4, productivity': (-10, 10),
            'feeder 5, level': (-5, 5),
            'feeder 5, speed': (-3, 3),
            'feeder 5, productivity': (-10, 10),
            'feeder 6, level': (-5, 5),
            'feeder 6, speed': (-3, 3),
            'feeder 6, productivity': (-10, 10),
            'feeder 7, speed': (-3, 3),
            'feeder 7, level': (-5, 5),
            'feeder 8, level': (-5, 5),
            'vacuum in the bunker': (-0.01, 0.01),
            'Overall blast volume, m3/h': (-100, 100),
            'natural gas flow': (-20, 20),
            'Oxygen content in the blast, %': (-1, 1),
            'blast furnace temperature': (-50, 50),
            'Temperature of exhaust gases in the off-gas duct, °C': (-50, 50),
            'temperature of the feed, matte siphon': (-50, 50),
            'temperature of the feed, melting zone, point 1': (-50, 50),
            'temperature of the feed, melting zone, point 2': (-50, 50),
            'temperature of natural gas': (-30, 30),
        }

        while True:  # Infinite loop to repeat data
            for row in data:
                try:
                    # Add noise to numeric fields
                    new_row = row.copy()
                    new_row['Date'] = datetime.now().isoformat()  # Update timestamp
                    for key, value in row.items():
                        if key in noise_ranges and isinstance(value, (int, float)):
                            noise = np.random.uniform(*noise_ranges[key])
                            new_row[key] = max(0, value + noise)  # Ensure non-negative values
                    await websocket.send(json.dumps(new_row))
                    await asyncio.sleep(2)  # 2s delay
                except websockets.exceptions.ConnectionClosed:
                    print("WebSocket connection closed, stopping simulation")
                    return
                    
    except Exception as e:
        print(f"Error in simulate_furnace: {e}")

async def handle_connection(websocket, path=None):
    print(f"New WebSocket connection, path: {path}")
    try:
        await simulate_furnace(websocket)
    except Exception as e:
        print(f"WebSocket error: {e}")
    finally:
        print("WebSocket connection closed")

async def main():
    server = await websockets.serve(handle_connection, "0.0.0.0", 5001)
    print("WebSocket server is running on ws://0.0.0.0:5001")
    await server.wait_closed()

if __name__ == "__main__":
    import json
    asyncio.run(main())